testing/fpgainfo/test_fpgainfo_c.cpp

```diff
@@ -0,0 +1,983 @@
+
+fpga_result errors_command(fpga_token *tokens, int num_tokens, int argc,
+			   char *argv[]);
+void errors_help(void);
+
+fpga_result fme_filter(fpga_properties *filter, int argc, char *argv[]);
+
+fpga_result fme_command(fpga_token *tokens, int num_tokens, int argc,
+			char *argv[]);
+void fme_help(void);
+
+void fpgainfo_print_common(const char *hdr, fpga_properties props);
+
+void fpgainfo_print_err(const char *s, fpga_result res);
+
+fpga_result port_filter(fpga_properties *filter, int argc, char *argv[]);
+
+fpga_result port_command(fpga_token *tokens, int num_tokens, int argc,
+			 char *argv[]);
+void port_help(void);
+
+fpga_result power_filter(fpga_properties *filter, int argc, char *argv[]);
+
+fpga_result power_command(fpga_token *tokens, int num_tokens, int argc,
+			  char *argv[]);
+void power_help(void);
+
+fpga_result temp_filter(fpga_properties *filter, int argc, char *argv[]);
+
+fpga_result temp_command(fpga_token *tokens, int num_tokens, int argc,
+			 char *argv[]);
+void temp_help(void);
+
+fpga_result bmc_filter(fpga_properties *filter, int argc, char *argv[]);
+
+fpga_result bmc_command(fpga_token *tokens, int num_tokens, int argc,
+			 char *argv[]);
+void bmc_help(void);
+
+fpga_result get_metrics(fpga_token token, metrics_inquiry inquiry,
+                        fpga_metric_info *metrics_info, fpga_metric *metrics,
+                        uint64_t *num_metrics);
+
+int fpgainfo_main(int argc, char *argv[]);
+
+}
+
+#include <iostream>
+#include <vector>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <errno.h>
+#include <unistd.h>
+#include "gtest/gtest.h"
+#include "test_system.h"
+
+using namespace opae::testing;
+
+class fpgainfo_c_p : public ::testing::TestWithParam<std::string> {
+    protected:
+        fpgainfo_c_p() {}
+
+        virtual void SetUp() override 
+        {
+            std::string platform_key = GetParam();
+            ASSERT_TRUE(test_platform::exists(platform_key));
+            platform_ = test_platform::get(platform_key);
+            system_ = test_system::instance();
+            system_->initialize();
+            system_->prepare_syfs(platform_);
+
+            EXPECT_EQ(fpgaInitialize(nullptr), FPGA_OK);
+
+            optind = 0;
+        }
+
+        virtual void TearDown() override {
+            system_->finalize();
+        }
+
+        test_platform platform_;
+        test_system *system_;
+};
+
+/**
+ * @test       help
+ * @brief      Test: help
+ * @details    help displays the application help message.<br>
+ */
+TEST_P(fpgainfo_c_p, help) {
+    help();
+}
+
+/**
+ * @test       parse_args0
+ * @brief      Test: parse_args
+ * @details    When passed with valid command options, <br>
+ *             the fn returns 0. <br>
+ */
+TEST_P(fpgainfo_c_p, parse_args0) {
+    char zero[20];
+    char one[20];
+    char *argv[] = { zero, one };
+
+    strcpy(zero, "fpgainfo");
+    strcpy(one, "fme");
+    EXPECT_EQ(parse_args(2, argv), 0);
```

Comment:
Scott must be setting optind back to 0 internally.  I would have suspected this to fail without explicitly resetting optind, but looks like it is passing. Good!